app/src/tracing.h

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "constants.h"
+
+#include <time.h>
+#include <regex>
+#include <sstream>
+#include <functional>
+#include <ccf/endpoint_context.h>
+#include <ccf/base_endpoint_registry.h>
+
+namespace scitt
+{
+  std::string REQUEST_ID_HEADER = "x-ms-request-id";
+  std::string CLIENT_REQUEST_ID_HEADER = "x-ms-client-request-id";
+
+  std::regex CLIENT_REQUEST_ID_REGEX("^[0-9a-zA-Z-]+");
+
+  thread_local std::string request_id;
+  thread_local std::optional<std::string> client_request_id;
+
+  int diff_timespec_ms(const struct timespec& time0, const struct timespec& time1) {
+    return (time1.tv_sec - time0.tv_sec) * 1000
+        + (time1.tv_nsec - time0.tv_nsec) / 1000000;
+  }
+
+  std::string create_request_id() {
+    std::stringstream stream;
+    stream << std::hex << rand();
+    return stream.str();
+  }
+
+  inline void SCITT_INFO(const std::string& msg)
+  {
+    CCF_APP_INFO("ClientRequestId={} RequestId={} {}",
+        client_request_id.value_or(""),
+        request_id,
+        msg
+        );
+  }
+
+  inline void SCITT_FAIL(const std::string& msg)
+  {
+    CCF_APP_FAIL("ClientRequestId={} RequestId={} {}",
+        client_request_id.value_or(""),
+        request_id,
+        msg
+        );
+  }
+
+  template <typename Fn, typename Ctx>
+  Fn generic_tracing_adapter(
+    Fn fn, const std::function<ccf::ApiResult(::timespec& time)>& get_time)
+  {
+    return [fn, get_time](Ctx& ctx) {
+      request_id = create_request_id();
```

Comment:
I think we should clear this at the end of the function. A try/finally is probably the safest way to do so.

One pattern I'm imagining is we run an endpoint, variable gets set, endpoint terminates, we run indexing strategy code (which isn't part of a request) on the same thread, it inherits the leftover request ID and that gets logs, which would be pretty confusing.

Reply:
Done. Good call.